drivers: phy: add new driver for davicom dm8806 phy

robi251 opened the pull request:
New driver for Davicom DM8806 PHY. Only the basic functionality. This is still a draft. please note, that these driver will only working if the related enhancement "Additional PHY API parameter" will be accepted:

pdgendt commented on drivers/ethernet/phy/Kconfig:
```diff
@@ -40,6 +40,12 @@ config PHY_ADIN2111
 	  Enable ADIN2111 PHY driver.
 
+config PHY_DM8806
+	bool "DM8806 PHY driver"
+	default y
```
You need to add `depends on DT_HAS_DAVICOM_DM8806_PHY_ENABLED`

robi251 replied:
Agree - dependency added

pdgendt commented:
Also please just force push existing PRs instead of creating new ones.

decsny commented on drivers/ethernet/phy/phy_dm8806.c:
```diff
@@ -0,0 +1,276 @@
+#define LOG_LEVEL       CONFIG_ETHERNET_LOG_LEVEL
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
```
you don't need to define macros for the arguments, also this macro has a form that can take both arguments

robi251 replied:
Useless definition are removed

pdgendt commented:
Squash your commits and force push to this PR, [fixup commits are not allowed](https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow).

robi251 committed:
drivers: dm8806: add new driver for davicom dm8806 phy mac

New driver for Davicom DM8806 PHY. Driver is using standar mdio API
to manage the DM8806 switch controller. Register access needs the
PHY addres or switch address to be one of five possible values, since
DM8806 has built-in five PHY's. These values should be defined in the
application .dts file. One DM8806 ethernet port must corresponds with
one ethernet PHY node with two properties for ethernet port: one for
PHY address and one for switch address - <reg> for register access from
Internal PHY Register area and <reg-switch> for register access from
Switch Per-Port Registers area. Device tree example below:

example device-tree:
  dm8806_phy: ethernet-phy@0 {
    reg = <2>;
    reg-switch = <8>;
    compatible = "davicom,dm8806-phy";
    status = "okay";
    davicom,interface-type = "rmii";
    reset-gpio = <&gpiod 2 GPIO_ACTIVE_LOW>;
    interrupt-gpio = <&gpioc 1 GPIO_ACTIVE_HIGH>;
  };

Signed-off-by: Robert Slawinski <robert.slawinski1@gmail.com>

robi251 commented:
@pdgendt @decsny HOw do You think about this implementation?

robi251 committed:
drivers: dm8806: link speed change interrupt handling

On the interrupt handling, one thread per driver instance is involved
into monitoring the semaphor, sends inside the gpio callback triggered
by the gpio interrupt. Each time, when the link parameters are change,
the DM8806 is generating the gpio interrupt. After getting semaphor,
the application callback function which was linked during initialization
process is called to get the new link parameters with standard API
calls

Signed-off-by: Robert Slawinski <robert.slawinski1@gmail.com>

pdgendt commented:
Please squash all your commits into a single one. It does not make sense to introduce a driver and update it in the same PR.

robi251 commented:
> Please squash all your commits into a single one. It does not make sense to introduce a driver and update it in the same PR.

Of course - I will sqush them to one commit. Give me time until today moon

pdgendt commented:
> > Please squash all your commits into a single one. It does not make sense to introduce a driver and update it in the same PR.
> 
> Of course - I will sqush them to one commit. Give me time until today moon

Actually you need to split the PHY API change into a separate commit, which might need some more discussion TBH.

robi251 commented:
> > > Please squash all your commits into a single one. It does not make sense to introduce a driver and update it in the same PR.
> > 
> > 
> > Of course - I will sqush them to one commit. Give me time until today moon
> 
> Actually you need to split the PHY API change into a separate commit, which might need some more discussion TBH.

Will it OK, if I will:
1. Remove the completely changes from the PHY API and move it to another one MRQ
2. Squash rest of commits to one commit

robi251 commented:
In separate MRQ:
https://github.com/zephyrproject-rtos/zephyr/pull/79714
we reach the conclusion, that reading MAC address table shouldn't be a part of PHY API, but rather DSI. Because of this reason I deleted it from the driver finaly.

decsny commented:
> New driver for Davicom DM8806 PHY. Only the basic functionality. This is still a draft. please note, that these driver will only working if the related enhancement "Additional PHY API parameter" will be accepted:

is this still all the case ?

robi251 commented:
> > New driver for Davicom DM8806 PHY. Only the basic functionality. This is still a draft. please note, that these driver will only working if the related enhancement "Additional PHY API parameter" will be accepted:
> 
> is this still all the case ?

Some update is required here:
1. This is now more than basic functionality. Driver is realized full phy API functionality
2. This driver will work only if: [PR81272](https://github.com/zephyrproject-rtos/zephyr/pull/81272) will be approved

Solution was tested on the real hardware with Cortex M4 MCU and is working stable. I don't know if there is more users who can run it on the HW. DM8806 is not so popular :(

robi251 commented:
Hi all.
One simple, but not simple question. I'm trying to figure out why  the pipeline complains because of `phy_dm8806_api`. It says that this API struct is defined but not used as well as the initialization function `phy_dm8806_init`. The construction is the same like in other drivers. What I've missed...?
@tbursztyka or @decsny  - I will be gretefull for any any tips, thank you

decsny commented:
I didn't look into it but might have something to do with the change to use DEVICE_API macro which was done recently

pdgendt commented on drivers/ethernet/phy/phy_dm8806.c:
```diff
@@ -0,0 +1,466 @@
+	return res;
+}
+
+static const struct ethphy_driver_api phy_dm8806_api = {
```
```suggestion
static DEVICE_API(ethphy, phy_dm8806_api) = {
```

robi251 replied:
here we go :) Thank you, now it works!

pdgendt reviewed (commented):
> I didn't look into it but might have something to do with the change to use DEVICE_API macro which was done recently

It shouldn't. That being said, you do need to use the wrapper macro. The warnings you get are if the `FOREACH` macro expands to nothing.

pdgendt commented on drivers/ethernet/phy/Kconfig.dm8806:
```diff
@@ -0,0 +1,49 @@
+menuconfig PHY_DM8806
+	bool "Davicom PHY DM8806 driver"
+	default y
+    depends on DT_HAS_DAVICOM_DM8806_PHY_ENABLED
```
Fix indent

robi251 replied:
Fixed!

fabiobaltieri commented on drivers/ethernet/phy/phy_dm8806.c:
```diff
@@ -0,0 +1,466 @@
+	static const struct phy_dm8806_config phy_dm8806_config_##n = {                            \
+		.mdio = DEVICE_DT_GET(DT_INST_BUS(n)),                                             \
+		.phy_addr = DT_INST_REG_ADDR(n),                                                   \
+		.switch_addr = DT_PROP(DT_NODELABEL(dm8806_phy##n), reg_switch),                   \
```
this also looks incorrect, it's making an assumption about a nodelabel, I'd imagine this should have been `DT_INST_PROP(n, reg_switch)`

kartben replied:
ya it's expecting the device node to have a label for some reason ...

```
dm8806_phy0: ethernet-phy@0 {
			reg = <0>;
			reg-switch = <8>
			...
```

kartben commented on drivers/ethernet/phy/phy_dm8806.c:
```diff
@@ -0,0 +1,466 @@
+		.mdio = DEVICE_DT_GET(DT_INST_BUS(n)),                                             \
+		.phy_addr = DT_INST_REG_ADDR(n),                                                   \
+		.switch_addr = DT_PROP(DT_NODELABEL(dm8806_phy##n), reg_switch),                   \
+		.gpio_int = GPIO_DT_SPEC_INST_GET(n, interrupt_gpio),                              \
+		.gpio_rst = GPIO_DT_SPEC_INST_GET(n, reset_gpio),                                  \
```
and this is incorrect too...

should be `int_gpios` and `reset_gpios`, plus, the properties are not required so it prob calls for a `GPIO_DT_SPEC_INST_GET_OR` or something ...

kartben commented:
@jukkar @pdgendt this driver seems very broken. I kind of managed to add it to build_all/ethernet and eventually get it to compile, but I wonder if it should be pulled out or something.

@robi251 Any chance you can contribute whatever improvements/fixes you've certainly had to implement if you're actually using this driver?

pdgendt commented:
> @jukkar @pdgendt this driver seems very broken. I kind of managed to add it to build_all/ethernet and eventually get it to compile, but I wonder if it should be pulled out or something.
> 
> @robi251 Any chance you can contribute whatever improvements/fixes you've certainly had to implement if you're actually using this driver?

I've opened a PR here https://github.com/zephyrproject-rtos/zephyr/pull/85925

This makes the driver build, no way to test it though. @robi251 can you test with the latest changes?
